File: app/src/main/java/com/kollectivemobile/euki/cycleig/CycleIgBundleBuilder.java
package com.kollectivemobile.euki.cycleig;

import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.kollectivemobile.euki.model.database.entity.CalendarItem;
import com.kollectivemobile.euki.utils.Constants;

import java.util.Date;
import java.util.List;

public class CycleIgBundleBuilder {
    private static final String CYCLE_CANONICAL = "https://cycle.fhir.me";
    private static final String CYCLE_CODE_SYSTEM = CYCLE_CANONICAL + "/CodeSystem/cycle";
    private static final String EUKI_CODE_SYSTEM = "https://github.com/Euki-Inc/Euki-Android/CodeSystem/euki-cycle-ig";
    private static final String SNOMED = "http://snomed.info/sct";
    private static final String OBSERVATION_CATEGORY = "http://terminology.hl7.org/CodeSystem/observation-category";
    private static final String BUNDLE_PROFILE = CYCLE_CANONICAL + "/StructureDefinition/period-tracking-bundle";
    private static final String FACT_PROFILE = CYCLE_CANONICAL + "/StructureDefinition/period-tracking-fact";
    private static final String MENSTRUAL_BLEEDING_PROFILE = CYCLE_CANONICAL + "/StructureDefinition/menstrual-bleeding";
    private static final String MENSTRUAL_FLOW_PROFILE = CYCLE_CANONICAL + "/StructureDefinition/menstrual-flow";
    private static final String SYMPTOM_PROFILE = CYCLE_CANONICAL + "/StructureDefinition/symptom";

    private int observationIndex = 0;
    private JsonArray entries = new JsonArray();

    public String build(CycleIgSnapshot snapshot) {
        observationIndex = 0;
        entries = new JsonArray();

        for (CalendarItem item : snapshot.getItems()) {
            addBleedingFacts(item);
            if (snapshot.getScope().includeEmotions()) {
                addEmotionFacts(item);
            }
            if (snapshot.getScope().includeBody()) {
                addBodyFacts(item);
            }
            if (snapshot.getScope().includeNotes()) {
                addNoteFact(item);
            }
        }

        if (observationIndex == 0) {
            throw new IllegalArgumentException("SMART Link requires at least one recorded fact.");
        }

        JsonObject bundle = new JsonObject();
        bundle.addProperty("resourceType", "Bundle");
        bundle.addProperty("id", "euki-cycle-ig-export");
        JsonObject meta = new JsonObject();
        JsonArray profiles = new JsonArray();
        profiles.add(BUNDLE_PROFILE);
        meta.add("profile", profiles);
        bundle.add("meta", meta);
        bundle.addProperty("type", "collection");
        bundle.addProperty("timestamp", new Date().toInstant().toString());
        bundle.add("entry", entries);
        return new GsonBuilder().setPrettyPrinting().create().toJson(bundle);
    }

    private void addBleedingFacts(CalendarItem item) {
        if (!CycleIgSnapshotFactory.hasBleedingValue(item)) {
            return;
        }

        String date = CycleIgDates.formatDate(item.getDate());
        boolean menstrualBleeding = item.getIncludeCycleSummary();
        JsonObject bleeding = baseObservation(MENSTRUAL_BLEEDING_PROFILE, cycleCode("menstrual-bleeding", "Menstrual bleeding"), date);
        bleeding.addProperty("valueBoolean", menstrualBleeding);
        addObservation(bleeding);

        if (menstrualBleeding && item.getBleedingSize() != null) {
            Flow flow = flowFor(item.getBleedingSize());
            JsonObject flowObservation = baseObservation(MENSTRUAL_FLOW_PROFILE, cycleCode("menstrual-flow", "Patient-reported menstrual flow category"), date);
            flowObservation.add("valueCodeableConcept", codeableConcept(CYCLE_CODE_SYSTEM, flow.code, flow.display, flow.display.toLowerCase()));
            addObservation(flowObservation);
        }

        addCounterFacts(item, item.getBleedingProductsCounter(), Constants.BleedingProducts.values, "bleeding-product", "Bleeding product", date);
        addCounterFacts(item, item.getBleedingClotsCounter(), Constants.BleedingClots.values, "bleeding-clot", "Bleeding clot", date);
    }

    private void addCounterFacts(CalendarItem item, List<Integer> counters, Object[] values, String code, String display, String date) {
        if (counters == null) {
            return;
        }

        for (int index = 0; index < counters.size() && index < values.length; index++) {
            Integer count = counters.get(index);
            if (count == null || count <= 0) {
                continue;
            }

            String valueCode = code + "-" + (index + 1);
            String valueDisplay = values[index].toString().toLowerCase().replace('_', ' ');
            JsonObject observation = baseObservation(FACT_PROFILE, appCode(code, display), date);
            JsonObject value = codeableConcept(EUKI_CODE_SYSTEM, valueCode, valueDisplay, valueDisplay);
            value.addProperty("text", valueDisplay + " x" + count);
            observation.add("valueCodeableConcept", value);
            addObservation(observation);
        }
    }

    private void addEmotionFacts(CalendarItem item) {
        if (item.getEmotions() == null) {
            return;
        }

        for (Constants.Emotions emotion : item.getEmotions()) {
            addSymptom(item, "emotion." + emotion.name().toLowerCase(), label(emotion.name()), null, null);
        }
    }

    private void addBodyFacts(CalendarItem item) {
        if (item.getBody() == null) {
            return;
        }

        for (Constants.Body body : item.getBody()) {
            StandardCoding coding = standardBodyCoding(body);
            addSymptom(item, "body." + body.name().toLowerCase(), label(body.name()), coding, body == Constants.Body.SEVEREPAIN ? "Severe pain" : null);
        }
    }

    private void addNoteFact(CalendarItem item) {
        if (!item.hasNote()) {
            return;
        }

        JsonObject observation = baseObservation(FACT_PROFILE, appCode("daily-note", "Daily note"), CycleIgDates.formatDate(item.getDate()));
        observation.addProperty("valueString", item.getNote());
        addObservation(observation);
    }

    private void addSymptom(CalendarItem item, String code, String display, StandardCoding standardCoding, String text) {
        JsonObject observation = baseObservation(SYMPTOM_PROFILE, cycleCode("symptom", "Symptom"), CycleIgDates.formatDate(item.getDate()));
        JsonObject value = new JsonObject();
        JsonArray coding = new JsonArray();
        if (standardCoding != null) {
            coding.add(coding(standardCoding.system, standardCoding.code, standardCoding.display));
        }
        coding.add(coding(EUKI_CODE_SYSTEM, code, display));
        value.add("coding", coding);
        value.addProperty("text", text == null ? display : text);
        observation.add("valueCodeableConcept", value);
        addObservation(observation);
    }

    private JsonObject baseObservation(String profile, JsonObject code, String date) {
        JsonObject observation = new JsonObject();
        observation.addProperty("resourceType", "Observation");
        observation.addProperty("id", nextId());
        observation.addProperty("status", "final");
        JsonObject meta = new JsonObject();
        JsonArray profiles = new JsonArray();
        profiles.add(profile);
        meta.add("profile", profiles);
        observation.add("meta", meta);
        observation.add("category", surveyCategory());
        observation.add("code", code);
        observation.addProperty("effectiveDateTime", date);
        return observation;
    }

    private void addObservation(JsonObject observation) {
        JsonObject entry = new JsonObject();
        entry.addProperty("fullUrl", "urn:euki:cycle-ig:" + observation.get("id").getAsString());
        entry.add("resource", observation);
        entries.add(entry);
    }

    private String nextId() {
        observationIndex++;
        return String.format("obs-%04d", observationIndex);
    }

    private static JsonArray surveyCategory() {
        JsonArray category = new JsonArray();
        JsonObject item = new JsonObject();
        JsonArray coding = new JsonArray();
        coding.add(coding(OBSERVATION_CATEGORY, "survey", "Survey"));
        item.add("coding", coding);
        category.add(item);
        return category;
    }

    private static JsonObject cycleCode(String code, String display) {
        return codeableConcept(CYCLE_CODE_SYSTEM, code, display, display);
    }

    private static JsonObject appCode(String code, String display) {
        return codeableConcept(EUKI_CODE_SYSTEM, code, display, display);
    }

    private static JsonObject codeableConcept(String system, String code, String display, String text) {
        JsonObject concept = new JsonObject();
        JsonArray coding = new JsonArray();
        coding.add(coding(system, code, display));
        concept.add("coding", coding);
        concept.addProperty("text", text);
        return concept;
    }

    private static JsonObject coding(String system, String code, String display) {
        JsonObject coding = new JsonObject();
        coding.addProperty("system", system);
        coding.addProperty("code", code);
        coding.addProperty("display", display);
        return coding;
    }

    private static Flow flowFor(Constants.BleedingSize bleedingSize) {
        return switch (bleedingSize) {
            case SPOTING -> new Flow("flow-spotting", "Spotting");
            case LIGHT -> new Flow("flow-light", "Light");
            case MEDIUM -> new Flow("flow-moderate", "Moderate");
            case HEAVY -> new Flow("flow-heavy", "Heavy");
        };
    }

    private static StandardCoding standardBodyCoding(Constants.Body body) {
        return switch (body) {
            case CRAMPS -> new StandardCoding(SNOMED, "431416001", "Menstrual cramp");
            case HEADACHE -> new StandardCoding(SNOMED, "25064002", "Headache");
            case FATIGUE -> new StandardCoding(SNOMED, "84229001", "Fatigue");
            case NAUSEAS -> new StandardCoding(SNOMED, "422587007", "Nausea");
            case TENDERBREASTS -> new StandardCoding(SNOMED, "55222007", "Breast tenderness");
            case FEVER -> new StandardCoding(SNOMED, "386661006", "Fever");
            case ACNE -> new StandardCoding(SNOMED, "11381005", "Acne");
            default -> null;
        };
    }

    private static String label(String name) {
        return name.toLowerCase().replace('_', ' ');
    }

    private static class Flow {
        final String code;
        final String display;

        Flow(String code, String display) {
            this.code = code;
            this.display = display;
        }
    }

    private static class StandardCoding {
        final String system;
        final String code;
        final String display;

        StandardCoding(String system, String code, String display) {
            this.system = system;
            this.code = code;
            this.display = display;
        }
    }
}

File: app/lib/cycle_ig/cycle_ig_bundle_builder.dart
import 'dart:convert';

import 'package:menstrudel/cycle_ig/cycle_ig_dates.dart';
import 'package:menstrudel/cycle_ig/cycle_ig_snapshot.dart';
import 'package:menstrudel/models/flows/flow_enum.dart';
import 'package:menstrudel/models/period_logs/pain_level_enum.dart';
import 'package:menstrudel/models/period_logs/log_day.dart';
import 'package:menstrudel/models/period_logs/symptom.dart';
import 'package:menstrudel/models/period_logs/symptom_type_enum.dart';

class CycleIgBundleBuilder {
  static const cycleSystem = 'https://cycle.fhir.me/CodeSystem/cycle';
  static const appSystem = 'https://menstrudel.app/fhir/CodeSystem/period-log';
  static const surveySystem =
      'http://terminology.hl7.org/CodeSystem/observation-category';

  static Map<String, dynamic> build(CycleIgSnapshot snapshot) {
    final entries = <Map<String, dynamic>>[];

    for (final log in snapshot.logs) {
      entries.add(_entry(_bleedingObservation(log)));

      if (snapshot.scope.includeFlow) {
        entries.add(_entry(_flowObservation(log)));
      }

      if (snapshot.scope.includeSymptoms) {
        for (final symptom in log.symptoms) {
          entries.add(_entry(_symptomObservation(log, symptom)));
        }
      }

      if (snapshot.scope.includePain && log.painLevel != null) {
        entries.add(_entry(_painLevelObservation(log)));
      }
    }

    return {
      'resourceType': 'Bundle',
      'meta': {
        'profile': [
          'https://cycle.fhir.me/StructureDefinition/period-tracking-bundle',
        ],
      },
      'type': 'collection',
      'timestamp': snapshot.generatedAt.toUtc().toIso8601String(),
      'entry': entries,
    };
  }

  static String encode(CycleIgSnapshot snapshot) => jsonEncode(build(snapshot));

  static Map<String, dynamic> _bleedingObservation(LogDay log) {
    final date = CycleIgDates.isoDate(log.date);
    return {
      'resourceType': 'Observation',
      'id': _id('menstrual-bleeding', date),
      'meta': {
        'profile': [
          'https://cycle.fhir.me/StructureDefinition/menstrual-bleeding',
        ],
      },
      'status': 'final',
      'category': [_surveyCategory()],
      'code': {
        'coding': [
          {
            'system': cycleSystem,
            'code': 'menstrual-bleeding',
            'display': 'Menstrual bleeding',
          },
        ],
      },
      'effectiveDateTime': date,
      'valueBoolean': log.flow != FlowRate.none,
    };
  }

  static Map<String, dynamic> _flowObservation(LogDay log) {
    final date = CycleIgDates.isoDate(log.date);
    final flowCode = _flowCode(log.flow);
    return {
      'resourceType': 'Observation',
      'id': _id('menstrual-flow', date),
      'meta': {
        'profile': ['https://cycle.fhir.me/StructureDefinition/menstrual-flow'],
      },
      'status': 'final',
      'category': [_surveyCategory()],
      'code': {
        'coding': [
          {
            'system': cycleSystem,
            'code': 'menstrual-flow',
            'display': 'Patient-reported menstrual flow category',
          },
        ],
      },
      'effectiveDateTime': date,
      'valueCodeableConcept': {
        'coding': [
          {
            'system': cycleSystem,
            'code': flowCode,
            'display': _flowDisplay(log.flow),
          },
        ],
      },
    };
  }

  static Map<String, dynamic> _symptomObservation(LogDay log, Symptom symptom) {
    final date = CycleIgDates.isoDate(log.date);
    final concept = _symptomConcept(symptom);
    return {
      'resourceType': 'Observation',
      'id': _id('symptom', date, concept['code'] as String),
      'meta': {
        'profile': ['https://cycle.fhir.me/StructureDefinition/symptom'],
      },
      'status': 'final',
      'category': [_surveyCategory()],
      'code': {
        'coding': [
          {'system': cycleSystem, 'code': 'symptom', 'display': 'Symptom'},
        ],
      },
      'effectiveDateTime': date,
      'valueCodeableConcept': {
        'coding': [
          {
            'system': concept['system'],
            'code': concept['code'],
            'display': concept['display'],
          },
        ],
        'text': concept['display'],
      },
    };
  }

  static Map<String, dynamic> _painLevelObservation(LogDay log) {
    final date = CycleIgDates.isoDate(log.date);
    final painLevel = PainLevel.values[log.painLevel as int];
    final code = 'pain-level-${painLevel.name}';
    return {
      'resourceType': 'Observation',
      'id': _id('pain-level', date),
      'meta': {
        'profile': [
          'https://cycle.fhir.me/StructureDefinition/period-tracking-fact',
        ],
      },
      'status': 'final',
      'category': [_surveyCategory()],
      'code': {
        'coding': [
          {
            'system': appSystem,
            'code': 'pain-level',
            'display': 'Menstrudel pain level',
          },
        ],
        'text': 'Menstrudel pain level',
      },
      'effectiveDateTime': date,
      'valueCodeableConcept': {
        'coding': [
          {
            'system': appSystem,
            'code': code,
            'display': _title(painLevel.name),
          },
        ],
        'text': _title(painLevel.name),
      },
    };
  }

  static Map<String, dynamic> _surveyCategory() {
    return {
      'coding': [
        {'system': surveySystem, 'code': 'survey', 'display': 'Survey'},
      ],
    };
  }

  static Map<String, dynamic> _entry(Map<String, dynamic> resource) {
    return {
      'fullUrl': 'urn:menstrudel:${resource['id']}',
      'resource': resource,
    };
  }

  static String _flowCode(FlowRate flow) {
    return switch (flow) {
      FlowRate.none => 'flow-none',
      FlowRate.spotting => 'flow-spotting',
      FlowRate.light => 'flow-light',
      FlowRate.medium => 'flow-moderate',
      FlowRate.heavy => 'flow-heavy',
    };
  }

  static String _flowDisplay(FlowRate flow) {
    return switch (flow) {
      FlowRate.none => 'None',
      FlowRate.spotting => 'Spotting',
      FlowRate.light => 'Light',
      FlowRate.medium => 'Moderate',
      FlowRate.heavy => 'Heavy',
    };
  }

  static Map<String, String> _symptomConcept(Symptom symptom) {
    final standard = switch (symptom.type) {
      SymptomType.cramps => ('431416001', 'Menstrual cramp'),
      SymptomType.headache => ('25064002', 'Headache'),
      SymptomType.fatigue => ('84229001', 'Fatigue'),
      SymptomType.bloating => ('116289008', 'Abdominal bloating'),
      SymptomType.depressed => ('366979004', 'Depressed mood'),
      _ => null,
    };

    if (standard != null) {
      return {
        'system': 'http://snomed.info/sct',
        'code': standard.$1,
        'display': standard.$2,
      };
    }

    final source = symptom.type == SymptomType.custom
        ? symptom.customName
        : symptom.getDbName();
    return {
      'system': appSystem,
      'code': 'symptom-${_slug(source)}',
      'display': _title(source),
    };
  }

  static String _id(String prefix, String date, [String? suffix]) {
    final id = [prefix, date, if (suffix != null) _slug(suffix)].join('-');
    return id.length <= 64 ? id : id.substring(0, 64);
  }

  static String _slug(String value) {
    final slug = value
        .toLowerCase()
        .replaceAll(RegExp(r'[^a-z0-9]+'), '-')
        .replaceAll(RegExp(r'^-+|-+$'), '');
    return slug.isEmpty ? 'value' : slug;
  }

  static String _title(String value) {
    return value
        .split(RegExp(r'[\s_-]+'))
        .where((part) => part.isNotEmpty)
        .map((part) => part[0].toUpperCase() + part.substring(1))
        .join(' ');
  }
}
